### DISCUSSION TITLE
Added: LDM Notes for September 13th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5183] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-13.md

## Agenda

1. Feedback on static abstracts in interfaces

**Comment by HaloFour** [comment]
I'm a little surprised that a form of the `concrete` constraint discussion didn't come up.  To me, not being able to use the interface itself as the generic type argument is one of the weirder aspects of `static abstract`s, especially since it would seem to result in a `TypeLoadException` at runtime vs. a compiler error.

**Reply by 333fred** [reply]
@tannergooding, care to comment on that?

**Reply by Joe4evr** [reply]
For the record: a `concrete` constraint would have some use outside of static abstract members.

**Reply by tannergooding** [reply]
This is already blocked at the language level because `ISomeInterface<T>` isn't convertible to `ISomeInterface<ISomeInterface<T>>`.
* `where T : IUnaryPlus<T>` means `T` must implement `IUnaryPlus<T>`. If `T` is itself `IUnaryPlus<T>`, then it must therefore implement `IUnaryPlus<IUnaryPlus<T>>`. Recursion is fun 😄 

You can of course do dangerous things via reflection or unsafe code. But that's always possible and there are many language concepts that don't have representation there.

**Reply by HaloFour** [reply]
I wasn't referring to that in the case of self type, but I do see that it seems that the language is already enforcing that the interface cannot be used as a generic type argument if it has `static abstract` members, so perhaps this issue was already discussed/addressed:

```
error CS9101: The interface 'IBar' cannot be used as type parameter 'T' in the generic type or method 'C.M<T>()'. The constraint interface 'IFoo' or its base interface has static abstract members.
```

**Reply by acaly** [reply]
A reproduction of the runtime exception (it's actually `System.Security.VerificationException`): [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEDMACdAmXARgHYBYAKAG8LtbcdYAXAUwCcAzCAY2ewEkadauTqiiANmwRgAZ0atujXABZsAWQAUASgDcg2gF8K++rgIBhADwAVAHzYA7gAs2va9hD8TwsXSzZzbRNRH18xawA6TV1guiMROOME2n8mNk4efgAxRQB7VgBPG0Jbb1iU1XNWZggWGzxbbUcXauxrAk9rQj1k7HiTf3RCSXwAiAAbcZyuRnyiuzLe0UHhlTHJjT5p2cKbewydgq1y7FCwugO5iKqaurttHvP40X7ewYJrF225j2y83b4pV6ZzEg0q1VqzHqjS0zVcbQ6XiWQhOohgzAcAWhDxOz0S5AGeDwi3O5gmU3+RUB13JGnRmM+zG+hW0MV68QMQA===).

**Reply by acaly** [reply]
I think a `System.Security.VerificationException` is an indication that the language and runtime do not agree on what is a verified program. Since the example program I provided is really invalid, I think there should be something to do for the compiler to generate an error. The runtime can't do much here.

I have mentioned this example for multiple times in at least 3 different issues/discussions, expressing the same idea "not being able to use the interface itself as the generic type argument is one of the weirder aspects of static abstract". However, I haven't got any feedback on how or whether this `VerificationException` problem would potentially be addressed in the next version of C# or runtime.

**Reply by HaloFour** [reply]
That's what I think these discussions are for.  It sounds like the compiler is attempting to catch and prevent the simple case, but more complicated cases can get through.

This feature is only going out as preview in C# 10 specifically to give people an opportunity to take it for a test drive.  How the feature evolves before (potentially) being released as a part of .NET 7/C# 11 will depend on that feedback.

**Comment by svick** [comment]
> We can solve issues like this by implementing a self type, which will also just be a generally-useful feature for the language. We're positive about the feature, but it is going to need a decent amount of design work. In particular, there are interfaces such as `IEquatable<T>` that have been in C# since C# 2.0, and we need to think about if we want to be able to apply such a constraint to those interfaces, how that would work, and what level of breaking change it would be.

I was curious if there were types that would be broken by this in the BCL and didn't find any there, but [I did find some in ClrMD](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSB0AMgJZQCOA3OugG4CGY2YApgM4QA2w2AvNutkHYAZmAD2AW2x1spbAEEADooAikuqXwBhCGBZRgaiRqj4A4s2DzWrZhIBGHYmwAUASgFCA7gAtmLbABCOnwASVYVAE8oOgliAGNPQVFJbG45EItgAFEAD0UxMGBmABMAFUjFVw80IWxff2Y080tQg39hOnjq/HkoSJdiXgA+WXwAOVimnj4AIlDssgg6YDpHZgADJFnsADJdsayLKH8E+TAAcwgJZgNWdwBtTABdIL5gGrrbDmZ47mAqGhhIVmF0fNgXPRGOkoEw2JwPklcEgAJwuD4UIA), like [`CLrArray`](https://github.com/microsoft/clrmd/blob/d11a6a3/src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs):

```c#
public struct ClrArray : IEquatable<ClrArray>, IEquatable<ClrObject>
```

**Reply by alrz** [reply]
I wouldn't consider the existing `IEquatable<T>` as a "type class" for equality. There's `IEqualityOperators`.

**Reply by acaly** [reply]
> I wouldn't consider the existing `IEquatable<T>` as a "type class" for equality. There's `IEqualityOperators`.

I think no matter whether and how we will have self type constraint, `IEquatable` and `IEqualityOperators` are 2 different type classes. `operator==` isn't identical to (and thus a replacement of) `Equals` method today, for example, in the case of NaN equality, so both would be necessary.
